### DISCUSSION TITLE
Locally averaged solution in the loss term.

### DISCUSSION THREAD

**Discussion by detzer0** [lululxvi/deepxde#1123] [General]
Dear Prof. @lululxvi  and other contributors, 

First of all, thank you for such an amazing PINN library with the open-source code.  
 
I would like to code a 1-D PINN problem with the additional term  $\overset{M}{\underset{j=1}{\sum}} (\tilde{u}(x_{j}) - g(x_{j}))^{2}$ in the loss function, where $\tilde{u}(x_{j}) =  \frac{1}{\delta} \overset{x_{j}+\delta/2}{\underset{x_{j}-\delta/2}{\int}} u(x) \ dx$ stands for the local average over the interval of length $\delta$ of the PINN solution $u$ at the point $x_{j}, \ 1 \leq j \leq M$ and $g \in \mathbb{R}^{M}$ is some data vector. 

My idea for implementing this loss term is the following: 

1. The integral for the averaging is discretized with the uniform quadrature rule  $\tilde{u}(x_{j}) \approx \frac{1}{\delta N} \overset{N}{\underset{i=1}{\sum}} u(x_{i,j}^{q})$. Hence for every point $x_{j}$, there are $N$ associated quadrature points  $x_{1,j}^{q},...,x_{N,j}^{q}$. 
2. I would like to build a matrix array of shape $(M,N)$ with the $j$-th row defined as $(u(x_{1,j}^{q}), ...,u( x_{N,j}^{q}) )$ and to compute the mean of the matrix  to get the averaged values for the rows. 

However, I do not understand how to feed some particular quadrature points (apart from the collocation points) to the network and how to evaluate my PINN solution $u$ at these points $x_{i,j}^{q}$ to build such an array. How can I build  such a matrix in the training process?  I have a feeling that then this additional loss term could be incorporated through PointSetOperatorBC, in a manner similar to inverse problems. 

Desirably, I would like to pass the points $x_{j}$ ,where I would like to average the solution, to the network, and to sample the quadrature points of interest around these points. Then I have to evaluate the neural network on these points to build the aforementioned array. In addition, I have to use the values $\tilde{u}(x_{j})$ in some other function which I use for my training and hence they should be available as the output of the network. 

Could you kindly provide a hint on how it could be done?

**Comment by lululxvi** [comment]
You can look at the examples of integral PDE, or fractional PDE.
